File: P81.jl
using DelimitedFiles

mat=readdlm("p081_matrix.txt", ',')

function lastmove(i::Int, j::Int, mat::Matrix)
    if i*j==1
        return 0
    elseif i==1
        return mat[i, j-1]
    elseif j==1
        return mat[i-1, j]
    else
        return min(mat[i-1, j], mat[i, j-1])
    end
end

function recursivepath(mat::Matrix)
    m,n= size(mat)
    for i in 1:m
        for j in 1:n
            mat[i,j]+=lastmove(i,j,mat)
        end
    end
end

mat= [131 673 234 103 18; 
201 96 342 965 150;
630 803 746 422 111;
537 699 497 121 956;
805 732 524 37 331]
recursivepath(mat)
println(mat[end,end])

File: P81.py
import numpy as np

with open('p081_matrix.txt') as f:
# with open('test.txt') as f:
    matrix = np.array([list(map(int, string.strip().split(',')))
                       for string in f.readlines()])
    print(matrix)

def greedy_sum(matrix):
    m = len(matrix)
    i, j = 0, 0
    count = matrix[i, j]
    while i + 1 < m and j + 1 < m:
        if matrix[i, j + 1] >= matrix[i + 1, j]:
            i += 1
        else:
            j += 1
        print(matrix[i,j])
        count += matrix[i, j]
    if i>j:
        while j<m-1:
            j+=1
            count += matrix[i, j]
    else:
        while i<m-1:
            i+=1
            count += matrix[i, j]
    print(i,j)
    return count

def recurive_sum(matrix):
    n, m = len(matrix), len(matrix[0,:])
    for i in range(n):
        for j in range(m):
            matrix[i,j]+=min(matrix[i-1,j],matrix[i,j-1]) if i*j>0 else (matrix[i-1,j] if i else (matrix[i,j-1] if j else 0))

print(recurive_sum(matrix))
